### PR TITLE
Verifiable RPC Client Support

### PR DESCRIPTION
**Pull request by epociask**
**Extended from https://github.com/near/nearcore/pull/8644**
_Author(s):_ @epociask

## Change Description 
**Verify Trie Path Subcommand**
- Embed `verify_trie_path` subcommand in `neard` CLI definition that takes in _account_id_, _public_key_, _state_root_, & _proof_path_ parameters
- Returns OS status code based on whether the path was able to be successfully verified; `0` if valid, `1` if invalid

**View Access Key**
- Added support for a boolean `include_proof` query parameter for JSON API `view_access_key` method
- Extended RPC querying functionality and AccessKey primitives to support `Option<Arc<Vec<8>>>` proof type
- Generate _proof_ ~(ie merkle path)~ using Trie iterator that's configured to remember all visited Trie nodes starting from root
- Modified RPC integration test to ensure that the prevalence of a proof value is returned within the query response

**View Access Key List**
- Added support for a boolean `include_proof` query parameter for JSON API `view_access_key_list` method
- Reworked internal view function logic to return `Vec<(PublicKey, AccessKeyView)>` instead of `Vec<(PublicKey, AccessKey)>` 
- Generate _proof_ using same logical methodology as `view_access_key`
- Modified RPC integration test to ensure that the prevalence of a proof value is returned within the query response

### Integration Testing
* Added integration test to asset a series of invariants about the returned proof from 5 arbitrarily constructed nodes using the existing pytesting framework 
* Added support for proof verification using subprocess calls to neard CLI and making invariant assertions about the resultant status code

#### View Access Key 
Using default locally generated genesis state, queried `view_access_key` for `test.near` _account_id_ and `ed25519:2ZcmxtePPHt5fPioTCYxpv1VrEZH7i8kyGYNM1WGcYy2` for _public_key_ with the following request body:
```
curl --location --request POST 'http://0.0.0.0:8332' \
--header 'Content-Type: application/json' \
--data-raw '{
  "jsonrpc": "2.0",
  "id": "1",
  "method": "query",
  "params": {
    "request_type": "view_access_key",
    "finality": "final",
    "account_id": "test.near",
    "public_key": "ed25519:2ZcmxtePPHt5fPioTCYxpv1VrEZH7i8kyGYNM1WGcYy2",
    "include_proof": true
  }
}'
```

Which generated the following response:
```
{
"jsonrpc": "2.0",
"result:{
   "block_hash":"353Mo1n9vL9HRCJM4Yr9G8vCiVr1t1amDXV5LVWYstVi",
   "block_height": 21129,
   "nonce": 0,
   "permission": "FullAccess",
   "proof": [
 [3,1,0,0,0,16,192,245,53,162,217,209,135,151,181,55,146,182,95,167,70,52,53,13,190,245,212,144,217,10,107,226,58,43,134,133,214,13,184,3,0,0,0,0,0,0],
[1,5,0,244,175,27,70,82,220,235,64,177,19,169,163,2,221,117,205,105,65,190,117,231,173,7,175,227,205,49,150,33,113,234,98,34,184,116,138,208,55,48,160,155,16,196,197,71,78,60,231,222,190,112,86,172,1,148,57,155,47,70,2,233,170,91,15,132,3,0,0,0,0,0,0],
[1,192,0,30,155,206,148,95,192,254,181,166,67,60,154,212,48,236,29,211,108,34,171,225,193,123,171,25,16,183,157,192,212,238,11,231,252,112,147,28,143,28,3,167,30,87,105,95,101,124,187,11,69,157,12,239,62,246,15,160,49,36,23,151,9,91,189,174,1,0,0,0,0,0,0],
[0,43,0,0,0,52,101,115,116,46,110,101,97,114,2,0,23,54,185,88,81,164,59,206,241,95,158,132,163,106,251,101,64,164,68,13,183,12,235,253,59,112,216,15,112,6,7,149,9,0,0,0,42,225,193,156,12,189,55,142,70,201,39,169,243,97,25,35,236,7,204,26,227,87,80,42,9,83,109,69,82,117,207,33,195,0,0,0,0,0,0,0]
]
},
"id":"1"
}
```

#### View Access Key List
Using default locally generated genesis state, queried `view_access_key_list` for `test.near` _account_id_ :
```
curl --location --request POST 'http://0.0.0.0:8332' \
--header 'Content-Type: application/json' \
--data-raw '{
  "jsonrpc": "2.0",
  "id": "1",
  "method": "query",
  "params": {
    "request_type": "view_access_key_list",
    "finality": "final",
    "account_id": "test.near",
    "include_proof": true
  }
}'
```

Which generated the following response:
```
{
    "jsonrpc": "2.0",
    "result": {
        "block_hash": "4LS8chXs3c67wX8vV9iSuZTDZEcMyKidQtZaDmAxTm8b",
        "block_height": 24737,
        "keys": [
            {
                "access_key": {
                    "nonce": 0,
                    "permission": "FullAccess",
                    "proof": [
                        [3,1,0,0,0,16,205,228,212,153,173,123,247,187,64,4,177,95,79,177,176,238,144,35,211,172,28,93,206,15,122,190,245,86,182,202,32,134,184,3,0,0,0,0,0,0],
                        [1,5,0,159,94,25,179,30,188,187,175,136,235,30,174,246,174,36,162,208,51,118,47,152,38,233,217,160,246,190,175,162,76,232,47,34,184,116,138,208,55,48,160,155,16,196,197,71,78,60,231,222,190,112,86,172,1,148,57,155,47,70,2,233,170,91,15,132,3,0,0,0,0,0,0],
                        [1,192,0,30,155,206,148,95,192,254,181,166,67,60,154,212,48,236,29,211,108,34,171,225,193,123,171,25,16,183,157,192,212,238,11,231,252,112,147,28,143,28,3,167,30,87,105,95,101,124,187,11,69,157,12,239,62,246,15,160,49,36,23,151,9,91,189,174,1,0,0,0,0,0,0],
                        [0,43,0,0,0,52,101,115,116,46,110,101,97,114,2,0,23,54,185,88,81,164,59,206,241,95,158,132,163,106,251,101,64,164,68,13,183,12,235,253,59,112,216,15,112,6,7,149,9,0,0,0,42,225,193,156,12,189,55,142,70,201,39,169,243,97,25,35,236,7,204,26,227,87,80,42,9,83,109,69,82,117,207,33,195,0,0,0,0,0,0,0]
                    ]
                },
                "public_key": "ed25519:2ZcmxtePPHt5fPioTCYxpv1VrEZH7i8kyGYNM1WGcYy2"
            }
        ]
    },
    "id": "1"
}
```

## Appendix Notes
**(03/05/2023)**
Validating inclusion of a state root, trie path, and arbitrary element turned out to be an incredibly non-trivial feat. At first I used the pre-provided Merkle tree logic and tried casting the nodes and root to merkle specific data structures which resulted in a failing inclusion check every time. Upon further investigation of the Trie implementation, it appears that hashes are being generated in a DFS manner with the resulting state root being irreflective of a merkle root. Because of this, validating a merkle path for the nearcore trie may require re-architecting the trie representation and flattening logic to support merkalization. 

**(03/09/2023)**
Verifying a trie path has turned out to be very very difficult. Only the pytesting integration testing framework provides assertive request/response interaction with a full node cluster that maintains Trie state. Other integration tests appear to stub the state interaction logic and default on mocked values. A CLI approach could be taken where the python tests make subprocess queries to a `neard` subcommand that performs adequate proof of inclusion within path verification. However, the relevant partial state construction [logic](https://github.com/near/nearcore/blob/master/core/store/src/trie/state_parts.rs#L182) uses a `Vec<[u8]>` [entry](https://github.com/near/nearcore/blob/ae8260d1446dde317e14634499cc2c94537c578b/core/primitives/src/challenge.rs#L9-L14) for what would be a proof node in the returned path. Since this variable is a statically defined array, its size must be known at compile time, making the procedure of converting a proof CLI input to a `PartialState` a seemingly complicated task. 

**(03/13/2023)**
Turns out that `serde` json library can read an arbitrary file into a `[u8]`

### CODE REVIEW

**Comment by nikurt** [comment]
Given this comment https://github.com/near/nearcore/pull/8644#issuecomment-1449279966, I'd like to reassign to the previous reviewers

**Comment by akhi3030** [comment]
Thanks nikurt!  CC: @mm-near

File: pytest/tests/sanity/verifiable_rpc_proofs.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        
+        key_query_result = node.get_access_key(node.signer_key.account_id,
+                                    node.signer_key.pk, proof=True)
+        
+        proof = key_query_result["result"]["proof"]
+        
+        logging.info("Fetched access key, validating for structural integrity....")
+
+        ## Invariant testing
+
```
Comment (mm-near) [inline pytest/tests/sanity/verifiable_rpc_proofs.py:48]: add a comment/TODO - that we're currently just checking a few invariants, and in the future, we should actually check that the proof hashes are correct.

File: pytest/tests/sanity/verifiable_rpc_proofs.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        branch_count = node_count(proof, BRANCH_NODE_NO_VALUE) + node_count(proof, BRANCH_NODE_WITH_VALUE)
+
+        assert leaf_count == 1; "There can not be more than one leaf node per proof"
+        assert branch_count >= 1; "There must be at least one branch node per proof"
+
+    logging.info("Testing complete, terminating nodes....")
+    for node in nodes:
+        node.kill()
+
+        # block_entry = node.get_latest_block()
```
Comment (mm-near) [inline pytest/tests/sanity/verifiable_rpc_proofs.py:67]: remove the commented out code

Comment (epociask) [reply pytest/tests/sanity/verifiable_rpc_proofs.py]: [2bcc159](https://github.com/near/nearcore/pull/8660/commits/2bcc159fa47942b865f143baa0e087d12f19d1d3)

File: pytest/tests/sanity/verifiable_rpc_proofs.py (hunk 1 above)
Comment (epociask) [reply pytest/tests/sanity/verifiable_rpc_proofs.py:47]: [2bcc159](https://github.com/near/nearcore/pull/8660/commits/2bcc159fa47942b865f143baa0e087d12f19d1d3)

File: pytest/tests/sanity/verifiable_rpc_proofs.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+BRANCH_NODE_NO_VALUE = 1
+BRANCH_NODE_WITH_VALUE = 2
+EXTENSION_NODE = 3
+
+node_encoding_keys = [LEAF_NODE, BRANCH_NODE_NO_VALUE, BRANCH_NODE_WITH_VALUE, EXTENSION_NODE]
+node_count = lambda proof, key: len([e for e in proof if e[0] == key])
+
+def main():
+    """
+        1. Spin up 5 nodes all using the same shard and iterate over each node 1 by 1
```
[the remark this answers is not in the document]
Comment (epociask) [reply pytest/tests/sanity/verifiable_rpc_proofs.py]: [2bcc159](https://github.com/near/nearcore/pull/8660/commits/2bcc159fa47942b865f143baa0e087d12f19d1d3)

File: runtime/runtime/src/state_viewer/mod.rs [hunk 4]
```diff
@@ -90,6 +110,7 @@ impl TrieViewer {
     ) -> Result<Vec<(PublicKey, AccessKey)>, errors::ViewAccessKeyError> {
         let prefix = trie_key_parsers::get_raw_prefix_for_access_keys(account_id);
         let raw_prefix: &[u8] = prefix.as_ref();
+
```
[the remark this answers is not in the document]
Comment (epociask) [reply runtime/runtime/src/state_viewer/mod.rs]: [2bcc159](https://github.com/near/nearcore/pull/8660/commits/2bcc159fa47942b865f143baa0e087d12f19d1d3)

File: runtime/runtime/src/state_viewer/mod.rs [hunk 5]
```diff
@@ -77,10 +77,30 @@ impl TrieViewer {
...
+        if include_proof {
+            let key = TrieKey::AccessKey {
+                account_id: account_id.clone(),
+                public_key: public_key.clone(),
+            }
+            .to_vec();
+
+            let mut iter = state_update.trie().iter()?;
+            iter.remember_visited_nodes(include_proof);
```
[the remark this answers is not in the document]
Comment (epociask) [reply runtime/runtime/src/state_viewer/mod.rs]: [2bcc159](https://github.com/near/nearcore/pull/8660/commits/2bcc159fa47942b865f143baa0e087d12f19d1d3)

File: runtime/runtime/src/state_viewer/mod.rs [hunk 6]
```diff
@@ -77,10 +77,30 @@ impl TrieViewer {
...
+        include_proof: bool,
+    ) -> Result<ViewAccessKeyResult, errors::ViewAccessKeyError> {
+        let access_key =
+            get_access_key(state_update, account_id, public_key)?.ok_or_else(|| {
+                errors::ViewAccessKeyError::AccessKeyDoesNotExist { public_key: public_key.clone() }
+            })?;
+
+        let mut view_result = ViewAccessKeyResult { access_key: access_key, proof: None };
+
+        if include_proof {
```
[the remark this answers is not in the document]
Comment (epociask) [reply runtime/runtime/src/state_viewer/mod.rs]: [2bcc159](https://github.com/near/nearcore/pull/8660/commits/2bcc159fa47942b865f143baa0e087d12f19d1d3)

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_query.rs [hunk 7]
```diff
@@ -222,7 +222,7 @@ fn test_query_access_keys() {
         let query_response = client
             .query(near_jsonrpc_primitives::types::query::RpcQueryRequest {
                 block_reference: BlockReference::latest(),
-                request: QueryRequest::ViewAccessKeyList { account_id: "test".parse().unwrap() },
+                request: QueryRequest::ViewAccessKeyList { account_id: "test".parse().unwrap()},
```
[the remark this answers is not in the document]
Comment (epociask) [reply chain/jsonrpc/jsonrpc-tests/tests/rpc_query.rs]: [2bcc159](https://github.com/near/nearcore/pull/8660/commits/2bcc159fa47942b865f143baa0e087d12f19d1d3)

File: neard/src/cli.rs [hunk 8]
```diff
@@ -670,6 +682,52 @@ pub enum VerifyProofError {
     InvalidBlockHashProof,
 }
 
+#[derive(clap::Parser)]
+pub struct VerifyTrieProofSubCommand {
```
Comment (mm-near) [inline neard/src/cli.rs:686]: mention in the comment, that this verifies that a given key belongs to a given account.

File: pytest/tests/sanity/verifiable_rpc_proofs.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        ##        as of now, we are just asserting a few invariants about the structural properties of a trie path
+
+        assert len(proof) > 0
+        "Valid proof returned must have more than one node"
+
+        for raw_node_encoding in proof:
+            assert raw_node_encoding[0] in node_encoding_keys
+            "0th bit of proof must be valid trie node encoding"
+
+        assert proof[-1][0] == LEAF_NODE
```
Comment (mm-near) [inline pytest/tests/sanity/verifiable_rpc_proofs.py:60]: btw - I think you're missing a comma in these assertions.

it should be:

assert x == y, "message"

(without the comma - the assert will not print the message - as python would treat it as separate no-op command)

**Review by mm-near — APPROVED** [review]

File: core/primitives/src/views.rs [hunk 10]
```diff
@@ -226,6 +237,13 @@ pub struct StateItem {
     pub proof: Vec<()>,
 }
 
+#[derive(serde::Serialize, serde::Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct ViewAccessKeyResult {
+    pub access_key: AccessKey,
+    // proof is a serialized array of nodes that make up a merkle path from the key to the root of the shard.
+    pub proof: Option<Vec<Arc<[u8]>>>,
```
Comment (bowenwang1996) [inline core/primitives/src/views.rs:244]: Why `Arc<[u8]>` here? If it is serialized it sounds like `Vec<u8>` would be more appropriate?

File: pytest/tests/sanity/verifiable_rpc_proofs.py (hunk 1 above)
Comment (bowenwang1996) [reply pytest/tests/sanity/verifiable_rpc_proofs.py:52]: @epociask do you plan to add verification of proofs in this test in the future?

**Comment by akhi3030** [comment]
This PR doesn't seem to be actively worked on.

**Pull request closed without merging**